File: sdk/go/sdk/buildkite/input_step.go
package buildkite

import "github.com/buildkite/buildkite-sdk/sdk/go/sdk/schema"

// Input Fields
type field struct {
	Text     *string                  `json:"text,omitempty"`
	Key      *string                  `json:"key,omitempty"`
	Hint     *string                  `json:"hint,omitempty"`
	Required *bool                    `json:"required,omitempty"`
	Default  *string                  `json:"default,omitempty"`
	Select   *string                  `json:"select,omitempty"`
	Options  []InputSelectFieldOption `json:"options,omitempty"`
	Multiple *bool                    `json:"multiple,omitempty"`
}

type Field interface {
	toSchema() field
}

type InputTextField struct {
	Text     *string
	Key      *string
	Hint     *string
	Required *bool
	Default  *string
}

func (i InputTextField) toSchema() field {
	field := field{
		Text:     i.Text,
		Key:      i.Key,
		Hint:     i.Hint,
		Required: i.Required,
		Default:  i.Default,
	}

	return field
}

type InputSelectFieldOption struct {
	Label string `json:"label,omitempty"`
	Value string `json:"value,omitempty"`
}

type InputSelectField struct {
	Select   *string
	Options  []InputSelectFieldOption
	Key      *string
	Hint     *string
	Required *bool
	Default  *string
	Multiple *bool
}

func (i InputSelectField) toSchema() field {
	field := field{
		Select:   i.Select,
		Hint:     i.Hint,
		Options:  i.Options,
		Key:      i.Key,
		Required: i.Required,
		Default:  i.Default,
		Multiple: i.Multiple,
	}

	return field
}

type InputStep struct {
	AllowDependencyFailure *bool
	Branches               []string
	DependsOn              DependsOn
	Fields                 []Field
	ID                     *string
	Identifier             *string
	If                     *string
	Input                  *string
	Key                    *string
	Label                  *string
	Name                   *string
	Prompt                 *string
}

func (step InputStep) ToPipelineStep() *PipelineStep {
	fields := make([]field, len(step.Fields))
	for i, field := range step.Fields {
		fields[i] = field.toSchema()
	}

	inputStep := &PipelineStep{
		Input: &schema.Input{
			String: step.Input,
		},

		Fields:     fields,
		ID:         step.ID,
		Identifier: step.Identifier,
		If:         step.If,
		Key:        step.Key,
		Label:      step.Label,
		Name:       step.Name,
		Prompt:     step.Prompt,
	}

	if step.AllowDependencyFailure != nil {
		inputStep.AllowDependencyFailure = &schema.AllowDependencyFailureUnion{
			Bool: step.AllowDependencyFailure,
		}
	}

	if step.Branches != nil {
		inputStep.Branches = &schema.Branches{
			StringArray: step.Branches,
		}
	}

	if step.DependsOn != nil {
		inputStep.DependsOn = step.DependsOn.toSchema()
	}

	return inputStep
}

File: sdk/csharp/src/Buildkite.Sdk/Schema/InputStep.cs
namespace Buildkite.Sdk.Schema;

/// <summary>
/// An input step collects information from the user before continuing.
/// </summary>
public class InputStep : IStep, IGroupStep
{
    /// <summary>The label for the input step.</summary>
    public string? Input { get; set; }

    /// <summary>Alias for Input.</summary>
    public string? Label { get; set; }

    /// <summary>Alias for Input.</summary>
    public string? Name { get; set; }

    /// <summary>A unique identifier for this step.</summary>
    public string? Key { get; set; }

    /// <summary>Alias for Key.</summary>
    public string? Id { get; set; }

    /// <summary>Alias for Key.</summary>
    public string? Identifier { get; set; }

    /// <summary>The message displayed in the input dialog.</summary>
    public string? Prompt { get; set; }

    /// <summary>Input fields to display.</summary>
    public List<Field>? Fields { get; set; }

    /// <summary>Branch filter pattern.</summary>
    public OneOrMany<string>? Branches { get; set; }

    /// <summary>A boolean expression to conditionally run this step.</summary>
    public string? If { get; set; }

    /// <summary>Step keys this step depends on.</summary>
    public OneOf<string, List<string>, List<Dependency>>? DependsOn { get; set; }

    /// <summary>Whether to proceed if a dependency fails.</summary>
    public bool? AllowDependencyFailure { get; set; }

    /// <summary>Team slugs or IDs permitted to provide input.</summary>
    public OneOrMany<string>? AllowedTeams { get; set; }

    /// <summary>The build state when waiting for input: 'passed', 'failed', or 'running'.</summary>
    public string? BlockedState { get; set; }

    /// <summary>The step type.</summary>
    public string Type => "input";
}
